File: src/main/go/leetcode/lists/hot100/LeetCode79.go
package hot100

func exist(board [][]byte, word string) bool {
	if len(board) == 0 {
		return true
	}
	visited := make([][]bool, len(board))
	for i := 0; i < len(board); i++ {
		visited[i] = make([]bool, len(board[0]))
	}
	var dfs func([][]byte, string, int, int, int, [][]bool) bool
	dfs = func(board [][]byte, word string, i int, j int, index int, visited [][]bool) bool {
		if index == len(word) {
			return true
		}
		if i < 0 || i > len(board)-1 || j < 0 || j > len(board[0])-1 ||
			board[i][j] != word[index] || visited[i][j] {
			return false
		}
		visited[i][j] = true
		if dfs(board, word, i+1, j, index+1, visited) ||
			dfs(board, word, i, j+1, index+1, visited) ||
			dfs(board, word, i-1, j, index+1, visited) ||
			dfs(board, word, i, j-1, index+1, visited) {
			return true
		}
		visited[i][j] = false
		return false
	}
	for i := 0; i < len(board); i++ {
		for j := 0; j < len(board)-1; j++ {
			if dfs(board, word, i, j, 0, visited) {
				return true
			}
		}
	}
	return false
}

File: src/main/java/leetcode/lists/hot100/LeetCode79.java
package leetcode.lists.hot100;

/**
 * @author aiwujingxin@gmail.com
 * @date 2022/9/11 19:11
 */
public class LeetCode79 {

    static boolean[][] visited;

    public boolean exist(char[][] board, String word) {

        if (board == null || board.length == 0) {
            return false;
        }
        visited = new boolean[board.length][board[0].length];

        for (int i = 0; i < board.length; i++) {
            for (int j = 0; j < board[0].length; j++) {
                if (dfs(board, word, i, j, 0)) {
                    return true;
                }
            }
        }
        return false;
    }

    private boolean dfs(char[][] board, String word, int i, int j, int index) {
        if (index == word.length()) {
            return true;
        }

        if (i < 0 || j < 0 || i > board.length - 1 || j > board[0].length - 1 || visited[i][j]
                || word.charAt(index) != board[i][j]) {
            return false;
        }
        visited[i][j] = true;
        if (dfs(board, word, i + 1, j, index + 1) ||
                dfs(board, word, i, j + 1, index + 1) ||
                dfs(board, word, i - 1, j, index + 1) ||
                dfs(board, word, i, j - 1, index + 1)) {
            return true;
        }
        visited[i][j] = false;
        return false;
    }
}
